File: src/test/resources/js-sdk-smoke/smoke.js
import fetch from 'node-fetch';
import PocketBase from 'pocketbase';
import EventSource from 'eventsource';

// Since Node 16+, global fetch is usually available, but we use node-fetch to be safe if running on older nodes
if (!globalThis.fetch) {
    globalThis.fetch = fetch;
}
globalThis.EventSource = EventSource;

async function run() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.error("Usage: node smoke.js <baseUrl>");
        process.exit(1);
    }

    const baseUrl = args[0];
    console.log(`Starting JS SDK smoke test against: ${baseUrl}`);

    const pb = new PocketBase(baseUrl);

    try {
        // 1. Create superuser via bootstrap (custom route)
        console.log("Bootstrapping superuser...");
        const bootstrapRes = await fetch(`${baseUrl}/api/bootstrap/superuser`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ email: 'smoke@example.com', password: 'password123' })
        });

        if (!bootstrapRes.ok && bootstrapRes.status !== 400) { // 400 if already bootstrapped
            throw new Error(`Failed to bootstrap: ${await bootstrapRes.text()}`);
        }

        // 2. Auth with password
        console.log("Authenticating...");
        const authData = await pb.collection('_superusers').authWithPassword('smoke@example.com', 'password123');
        console.log(`Authenticated as: ${authData.record.email}`);

        if (!pb.authStore.isValid) {
            throw new Error("SDK auth store rejected the returned auth token");
        }
        if (!pb.authStore.isSuperuser) {
            throw new Error("SDK auth store did not recognize the login as a superuser");
        }

        // 3. Testing auth refresh
        console.log("Testing auth refresh...");
        const refreshData = await pb.collection('_superusers').authRefresh();
        if (!pb.authStore.isValid) {
            throw new Error("Token refresh invalidated the authStore");
        }

        console.log("Testing logout...");
        pb.authStore.clear();
        if (pb.authStore.isValid) {
            throw new Error("AuthStore clear did not log out the user");
        }

        // Re-authenticate for subsequent steps
        await pb.collection('_superusers').authWithPassword('smoke@example.com', 'password123');

        // 4. Create a collection with rules allowing guest CRUD and including a file field
        console.log("Creating collection...");
        try {
            await pb.collections.create({
                name: "smoke_test_collection",
                type: "base",
                createRule: "",
                listRule: "",
                viewRule: "",
                updateRule: "",
                deleteRule: "",
                schema: [
                    { name: "title", type: "text", required: true },
                    { name: "count", type: "number" },
                    { name: "avatar", type: "file", options: { maxSelect: 1, maxSize: 5242880 } }
                ]
            });
        } catch (e) {
            // Ignore if already exists, but we update rules to be sure
            if (e.status !== 400) {
                throw e;
            }
        }

        // 5. Testing file upload & download & token
        console.log("Testing file upload...");
        const fileBlob = new Blob(["hello world"], { type: 'text/plain' });
        const fileRecord = await pb.collection('smoke_test_collection').create({
            title: "File Upload Test",
            count: 100,
            avatar: fileBlob
        });
        console.log(`File record created: ${fileRecord.id}, avatar: ${fileRecord.avatar}`);

        console.log("Testing file download URL...");
        const fileUrl = pb.files.getUrl(fileRecord, fileRecord.avatar);
        console.log(`Downloading from URL: ${fileUrl}`);
        const fileDlRes = await fetch(fileUrl);
        if (!fileDlRes.ok) {
            throw new Error(`Failed to download uploaded file: ${fileDlRes.statusText}`);
        }
        const fileText = await fileDlRes.text();
        if (fileText !== "hello world") {
            throw new Error(`File content mismatch: expected "hello world", got "${fileText}"`);
        }

        console.log("Testing protected file token...");
        const fileToken = await pb.files.getToken();
        if (!fileToken) {
            throw new Error("Failed to get protected file token");
        }

        // 6. Testing batch CRUD & rollback
        console.log("Testing batch create...");
        const batch = pb.createBatch();
        batch.collection('smoke_test_collection').create({ title: "Batch 1", count: 1 });
        batch.collection('smoke_test_collection').create({ title: "Batch 2", count: 2 });
        const batchResult = await batch.send();
        console.log("Batch Result is:", JSON.stringify(batchResult));
        if (!batchResult || !batchResult.responses || batchResult.responses.length !== 2) {
            throw new Error(`Batch send returned invalid responses, expected 2`);
        }

        console.log("Testing batch rollback on failure...");
        const beforeList = await pb.collection('smoke_test_collection').getList(1, 50);
        const beforeCount = beforeList.items.length;

        const failBatch = pb.createBatch();
        failBatch.collection('smoke_test_collection').create({ title: "Rollback Item", count: 1000 });
        failBatch.collection('non_existent_collection').create({ name: "error" });

        try {
            await failBatch.send();
            throw new Error("Batch with error should have failed");
        } catch (e) {
            console.log("Batch failed as expected, checking rollback...");
            const afterList = await pb.collection('smoke_test_collection').getList(1, 50);
            if (afterList.items.length !== beforeCount) {
                throw new Error(`Batch failure did not rollback. Records changed from ${beforeCount} to ${afterList.items.length}`);
            }
            console.log("Batch rollback verified!");
        }

        // 7. Testing realtime SSE
        console.log("Testing realtime SSE...");
        let realtimeEventReceived = false;
        await pb.collection('smoke_test_collection').subscribe('*', (e) => {
            console.log(`[Realtime Event] Action: ${e.action}, Record: ${e.record.id}`);
            if (e.action === 'create' && e.record.title === "Realtime Test") {
                realtimeEventReceived = true;
            }
        });

        await pb.collection('smoke_test_collection').create({
            title: "Realtime Test",
            count: 999
        });

        // Wait for event delivery
        for (let i = 0; i < 30; i++) {
            if (realtimeEventReceived) {
                break;
            }
            await new Promise(r => setTimeout(r, 100));
        }

        if (!realtimeEventReceived) {
            throw new Error("Realtime SSE event not received");
        }
        console.log("Realtime SSE event verified!");
        pb.collection('smoke_test_collection').unsubscribe('*');

        // Cleanup
        console.log("Cleaning up created records...");
        const finalCheckList = await pb.collection('smoke_test_collection').getList(1, 50);
        for (const item of finalCheckList.items) {
            await pb.collection('smoke_test_collection').delete(item.id);
        }

        console.log("✅ JS SDK Smoke Test Passed!");
        process.exit(0);

    } catch (err) {
        console.error("❌ JS SDK Smoke Test Failed:", err);
        process.exit(1);
    }
}

run();

File: src/test/resources/dart-sdk-smoke/smoke.dart
import 'dart:async';

import 'package:http/http.dart' as http;
import 'package:pocketbase/pocketbase.dart';

Future<void> main(List<String> args) async {
  if (args.isEmpty) {
    throw StateError('Usage: dart run smoke.dart <baseUrl>');
  }

  final pb = PocketBase(args.first);
  await pb.send('/api/bootstrap/superuser', method: 'POST', body: {
    'email': 'root@example.com',
    'password': 'secret123',
  });
  await pb.collection('_superusers').authWithPassword('root@example.com', 'secret123');
  await pb.collection('_superusers').authRefresh();
  if (!pb.authStore.isValid) {
    throw StateError('Dart SDK auth store rejected the refreshed token');
  }

  final collectionName = 'dart_smoke_posts';
  await pb.send('/api/collections', method: 'POST', body: {
    'name': collectionName,
    'type': 'base',
    'listRule': '',
    'viewRule': '',
    'createRule': '',
    'updateRule': '',
    'deleteRule': '',
    'fields': [
      {'name': 'title', 'type': 'text', 'required': true},
      {'name': 'count', 'type': 'number'},
      {'name': 'attachment', 'type': 'file'},
    ],
  });

  final created = await pb.collection(collectionName).create(body: {
    'title': 'from dart',
    'count': 1,
  });
  final updated = await pb.collection(collectionName).update(created.id, body: {
    'title': 'from dart updated',
    'count': 2,
  });
  if (updated.get<String>('title') != 'from dart updated') {
    throw StateError('record update did not roundtrip through Dart SDK');
  }

  final list = await pb.collection(collectionName).getList(page: 1, perPage: 10);
  if (list.items.isEmpty) {
    throw StateError('Dart SDK list returned no records');
  }

  final fileRecord = await pb.collection(collectionName).create(
    body: {'title': 'dart file', 'count': 3},
    files: [
      http.MultipartFile.fromString(
        'attachment',
        'hello from dart',
        filename: 'dart-smoke.txt',
      ),
    ],
  );
  final filename = fileRecord.get<String>('attachment');
  final fileResponse = await http.get(pb.files.getURL(fileRecord, filename));
  if (fileResponse.statusCode != 200 || fileResponse.body != 'hello from dart') {
    throw StateError('Dart SDK file upload/download did not roundtrip');
  }
  if ((await pb.files.getToken()).isEmpty) {
    throw StateError('Dart SDK did not receive a protected file token');
  }

  final batch = pb.createBatch()
    ..collection(collectionName).create(body: {
      'title': 'dart batch one',
      'count': 4,
    })
    ..collection(collectionName).create(body: {
      'title': 'dart batch two',
      'count': 5,
    });
  final batchResult = await batch.send();
  if (batchResult.length != 2 || batchResult.any((item) => item.status >= 400)) {
    throw StateError('Dart SDK batch response did not match the official array contract');
  }

  final realtimeEvent = Completer<RecordSubscriptionEvent>();
  final unsubscribe = await pb.collection(collectionName).subscribe('*', (event) {
    if (!realtimeEvent.isCompleted &&
        event.action == 'create' &&
        event.record?.get<String>('title') == 'dart realtime') {
      realtimeEvent.complete(event);
    }
  });
  final realtimeRecord = await pb.collection(collectionName).create(body: {
    'title': 'dart realtime',
    'count': 6,
  });
  final event = await realtimeEvent.future.timeout(const Duration(seconds: 10));
  if (event.record?.id != realtimeRecord.id) {
    throw StateError('Dart SDK realtime event returned the wrong record');
  }
  await unsubscribe();

  await pb.collection(collectionName).delete(created.id);
  await pb.collection(collectionName).delete(fileRecord.id);
  for (final item in await pb.collection(collectionName).getFullList()) {
    await pb.collection(collectionName).delete(item.id);
  }
  print('Dart SDK Smoke Test Passed!');
}
